### PR TITLE
Change the Notification permission  flow

### LINKED ISSUE #1095: [FR]: Notification permission dialog should not be shown immediately on first run
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Describe the problem

After running the app for the first time, the notification permission dialog is shown immediately. This is not a great user experience because the user doesn't have enough context to know what notifications are going to be shown, or how frequently. 

### Describe the solution

The notification permission dialog should only be shown after the user has completed the onboarding (i.e. chosen topics they are interested in) and there should be some additional information shown saying something like: 

"Would you like to be informed when new content is published on these topics, content is typically published every few days? [Yes/No]" 

If yes, show the notification permission dialog. 

### Additional context

_No response_

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

### PR DESCRIPTION
**Pull request by arvindri2005**
Fixes #1095 
Now Notification permission will be asked after knowing the interests of the user

https://github.com/android/nowinandroid/assets/70971577/092c8830-6d21-4f22-8a14-7737f6fa7124

### CODE REVIEW

File: feature/foryou/src/main/kotlin/com/google/samples/apps/nowinandroid/feature/foryou/ForYouScreen.kt [hunk 1]
```diff
@@ -444,6 +473,38 @@ fun TopicIcon(
...
+    onDismissRequest: ()-> Unit,
+    askNotificationPermission: () -> Unit
+) {
+    AlertDialog(
+        onDismissRequest = onDismissRequest,
+        title = {
+            Text(text = "Grant Notification Permission")
+        },
+        text = {
+            Text("Would you like to be informed when new content is published on these topics, content is typically published every few days?")
```
Comment (lihenggui) [inline feature/foryou/src/main/kotlin/com/google/samples/apps/nowinandroid/feature/foryou/ForYouScreen.kt:487]: String resources should be placed in `strings.xml`

File: feature/foryou/src/main/kotlin/com/google/samples/apps/nowinandroid/feature/foryou/ForYouViewModel.kt [hunk 2]
```diff
@@ -109,6 +110,23 @@ class ForYouViewModel @Inject constructor(
...
 
+    private val _showNotificationPermission = MutableStateFlow(false)
+    val showNotificationPermission: StateFlow<Boolean> = _showNotificationPermission
+    private val _dialogState = MutableStateFlow(false)
+    val dialogState: StateFlow<Boolean> = _dialogState
+
+    fun toggleDialog() {
+        viewModelScope.launch {
+            _dialogState.emit(!_dialogState.value)
+        }
```
Comment (hoc081098) [inline feature/foryou/src/main/kotlin/com/google/samples/apps/nowinandroid/feature/foryou/ForYouViewModel.kt:121]: _dialogState.update {} is simpler

File: feature/foryou/src/main/kotlin/com/google/samples/apps/nowinandroid/feature/foryou/ForYouViewModel.kt [hunk 3]
```diff
@@ -109,6 +110,23 @@ class ForYouViewModel @Inject constructor(
...
+    }
+    fun showNotificationPermission() {
+        viewModelScope.launch {
+            _showNotificationPermission.emit(!_showNotificationPermission.value)
+        }
```
Comment (hoc081098) [inline feature/foryou/src/main/kotlin/com/google/samples/apps/nowinandroid/feature/foryou/ForYouViewModel.kt:126]: Same as above 🙏

File: feature/foryou/src/main/kotlin/com/google/samples/apps/nowinandroid/feature/foryou/ForYouViewModel.kt (hunk 2 above)
Comment (arvindri2005) [reply feature/foryou/src/main/kotlin/com/google/samples/apps/nowinandroid/feature/foryou/ForYouViewModel.kt:121]: val showNotificationPermission = savedStateHandle.getStateFlow(SHOW_NOTIFICATION_PERMISSION, false)
    val dialogState = savedStateHandle.getStateFlow(DIALOG_STATE, false)
    fun toggleDialog() {
        savedStateHandle[SHOW_NOTIFICATION_PERMISSION] = !dialogState.value
    }
    fun showNotificationPermission() {
        savedStateHandle[DIALOG_STATE] = !showNotificationPermission.value
    }



you want something Like this

File: feature/foryou/src/main/kotlin/com/google/samples/apps/nowinandroid/feature/foryou/ForYouScreen.kt (hunk 1 above)
Comment (dturner) [reply feature/foryou/src/main/kotlin/com/google/samples/apps/nowinandroid/feature/foryou/ForYouScreen.kt:487]: @arvindri2005 Please do not resolve comments before they have actually been resolved.

Comment (arvindri2005) [reply feature/foryou/src/main/kotlin/com/google/samples/apps/nowinandroid/feature/foryou/ForYouScreen.kt]: I have fixed this in the latest commit cfa1d51

File: feature/foryou/src/main/kotlin/com/google/samples/apps/nowinandroid/feature/foryou/ForYouScreen.kt [hunk 4]
```diff
@@ -444,6 +473,38 @@ fun TopicIcon(
...
+                    askNotificationPermission()
+                    onDismissRequest()
+                },
+            ) {
+                Text(text = "OK")
+            }
+        },
+        dismissButton = {
+            NiaButton(onClick =  onDismissRequest ) {
+                Text(text ="Cancel")
```
Comment (jawidMuhammadi) [inline feature/foryou/src/main/kotlin/com/google/samples/apps/nowinandroid/feature/foryou/ForYouScreen.kt:501]: To maintain consistency and ease of localization, the "Ok" and the "Cancel" strings also should be included in the string file.

File: feature/foryou/src/main/kotlin/com/google/samples/apps/nowinandroid/feature/foryou/ForYouScreen.kt [hunk 5]
```diff
@@ -444,6 +473,38 @@ fun TopicIcon(
     )
 }
 
+@Composable
+private fun ShowDialog(
```
Comment (jawidMuhammadi) [inline feature/foryou/src/main/kotlin/com/google/samples/apps/nowinandroid/feature/foryou/ForYouScreen.kt:477]: I suggest renaming this function to `ShowNotificationPermissionDialog` for better clarity.

File: feature/foryou/src/main/kotlin/com/google/samples/apps/nowinandroid/feature/foryou/ForYouScreen.kt [hunk 6]
```diff
@@ -138,10 +146,14 @@ internal fun ForYouScreen(
...
     feedState: NewsFeedUiState,
     deepLinkedUserNewsResource: UserNewsResource?,
+    askShowDialog: Boolean,
+    askNotificationPermission: Boolean,
     onTopicCheckedChanged: (String, Boolean) -> Unit,
     onTopicClick: (String) -> Unit,
     onDeepLinkOpened: (String) -> Unit,
     saveFollowedTopics: () -> Unit,
+    showNotificationPermissionDialog: () -> Unit,
+    showNotificationPermission: () -> Unit,
```
Comment (jawidMuhammadi) [inline feature/foryou/src/main/kotlin/com/google/samples/apps/nowinandroid/feature/foryou/ForYouScreen.kt:156]: I suggest renaming these to `onNotificationPermissionDismiss` and `onNotificationPermissionConfirm` alternatively

File: feature/foryou/src/androidTest/kotlin/com/google/samples/apps/nowinandroid/feature/foryou/ForYouScreenTest.kt
Comment (alexvanyo) [inline feature/foryou/src/androidTest/kotlin/com/google/samples/apps/nowinandroid/feature/foryou/ForYouScreenTest.kt:1]: It looks like all of the changes here are to make existing tests pass, could we add a new test for the updated flow?

File: feature/foryou/src/main/kotlin/com/google/samples/apps/nowinandroid/feature/foryou/ForYouScreen.kt [hunk 7]
```diff
@@ -116,16 +118,22 @@ internal fun ForYouRoute(
...
         onboardingUiState = onboardingUiState,
         feedState = feedState,
         deepLinkedUserNewsResource = deepLinkedUserNewsResource,
+        askToShowNotificationPermissionDialog = notificationPermissionDialog,
+        askNotificationPermission = askNotificationPermission,
         onTopicCheckedChanged = viewModel::updateTopicSelection,
         onDeepLinkOpened = viewModel::onDeepLinkOpened,
         onTopicClick = onTopicClick,
         saveFollowedTopics = viewModel::dismissOnboarding,
+        onNotificationPermissionDismiss = viewModel::toggleNotificationPermissionDialog,
```
Comment (alexvanyo) [inline feature/foryou/src/main/kotlin/com/google/samples/apps/nowinandroid/feature/foryou/ForYouScreen.kt:135]: Similar to the comment in the viewmodel, this parameter is named `dismiss`, but it is used as a `toggle` (and called that way later). Could we change this to be two different functions to make the behavior more clear?

File: feature/foryou/src/main/kotlin/com/google/samples/apps/nowinandroid/feature/foryou/ForYouScreen.kt [hunk 8]
```diff
@@ -160,6 +172,18 @@ internal fun ForYouScreen(
...
+    if(askToShowNotificationPermissionDialog){
+        ShowNotificationPermissionDialog(
+            onNotificationPermissionDismiss,
+            onNotificationPermissionConfirm
+        )
+    }
+
+    if(askNotificationPermission){
+        NotificationPermissionEffect()
+        onNotificationPermissionConfirm()
```
Comment (alexvanyo) [inline feature/foryou/src/main/kotlin/com/google/samples/apps/nowinandroid/feature/foryou/ForYouScreen.kt:184]: This is a side-effect causing function called directly in composition - this should likely be wrapped in a `SideEffect` or something similar.

**Review by alexvanyo — CHANGES_REQUESTED** [review]

**Comment by keyboardsurfer** [comment]
Thank you for submitting this pull request. We have decided not to accept it at this time.
Feel free to comment or open an issue for further discussion if necessary.

**Pull request closed without merging**